File: MPP-firma-transport-java/FirmaServer/src/main/java/org/example/FirmaServices.java
package org.example;

import org.example.Exception.ValidationException;
import org.example.Validator.ValidatorRezervare;

import javax.crypto.Cipher;
import javax.crypto.spec.SecretKeySpec;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class FirmaServices implements IFirmaServices{

    private OficiuRepository oficiuRepo;
    private LocuriRepository locuriRepo;
    private CursaRepository cursaRepo;
    private RezervareRepository rezervareRepo;
    private Map<String,IFirmaObserver> loggedClients;

    private final int defaultThreadsNo=5;

    public FirmaServices(OficiuRepository oficiuRepo, LocuriRepository locuriRepo, CursaRepository cursaRepo, RezervareRepository rezervareRepo) {
        this.oficiuRepo = oficiuRepo;
        this.locuriRepo = locuriRepo;
        this.cursaRepo = cursaRepo;
        this.rezervareRepo = rezervareRepo;
        this.loggedClients = new ConcurrentHashMap<>();
    }

    @Override
    public synchronized Optional<Oficiu> login(String username, String password, IFirmaObserver client) throws AppException {
        //get oficiuDTO by username and password
        Optional<Oficiu> utilizatorOpt = oficiuRepo.findByUsernameAndPassword(username, password);



        if (utilizatorOpt.isPresent()) {
            //print the oficiu
            System.out.println(utilizatorOpt.get());
            Oficiu utilizator = utilizatorOpt.get();


                if (loggedClients.get(username) != null)
                    throw new AppException("User already logged in");

                loggedClients.put(username, client);
                return Optional.of(utilizator);

        }
        return Optional.empty();
    }

    @Override
    public synchronized void logout(String username) throws AppException {
        IFirmaObserver localClient=loggedClients.remove(username);
        if (localClient==null)
            throw new AppException("User "+username+" is not logged in.");
    }

    private String encrypt(String strToEncrypt, String secret) throws AppException {
        SecretKeySpec secretKey = new SecretKeySpec(secret.getBytes(), "AES");
        try {
            Cipher cipher = Cipher.getInstance("AES/ECB/PKCS5Padding");
            cipher.init(Cipher.ENCRYPT_MODE, secretKey);
            byte[] encryptedBytes = cipher.doFinal(strToEncrypt.getBytes());
            return Base64.getEncoder().encodeToString(encryptedBytes);
        }
        catch (Exception e){
            throw new AppException("Invalid password");
        }
    }

    private String decrypt(String strToDecrypt, String secret) throws AppException {
        SecretKeySpec secretKey = new SecretKeySpec(secret.getBytes(), "AES");
        try {
            Cipher cipher = Cipher.getInstance("AES/ECB/PKCS5Padding");
            cipher.init(Cipher.DECRYPT_MODE, secretKey);
            byte[] decryptedBytes = cipher.doFinal(Base64.getDecoder().decode(strToDecrypt));
            return new String(decryptedBytes);
        }
        catch (Exception e){
            throw new AppException("Invalid password");
        }
    }

    @Override
    public List<CursaDTO> getAllCurse() throws AppException {
        //get all curse and transform them into DTOs
        return cursaRepo.findAll().stream()
                .map(cursa -> {
                    var locuriDisponibile = 18-locuriRepo.findUltimulLoc(cursa.getId());
                    return new CursaDTO(cursa.getId(), cursa.getDestinatie(), cursa.getData(), locuriDisponibile);
                })
                .toList();
    }

    @Override
    public List<CursaDTO> getCursaByDestinatieData(String destinatie,String data) throws AppException {
        Optional<Cursa> optionalCursa = cursaRepo.findByDestinatieData(destinatie, data);

        if (optionalCursa.isPresent()) {
            Cursa cursa = optionalCursa.get();
            int locuriDisponibile = 18 - locuriRepo.findUltimulLoc(cursa.getId());
            CursaDTO cursaDTO = new CursaDTO(cursa.getId(), cursa.getDestinatie(), cursa.getData(), locuriDisponibile);
            return List.of(cursaDTO); // Return a list containing the DTO
        } else {
            return Collections.emptyList(); // Return an empty list if no matching record was found
        }
    }

    @Override
    public List<LocDTO> getLocuriDisponibile(Long idCursa) throws AppException {
        return locuriRepo.findByCursa(idCursa);
    }

    public RezervareDTO getRezervareById(Long rezervare) {
        return rezervareRepo.findOne(rezervare).map(rezervare1 -> new RezervareDTO(rezervare1.getId(), rezervare1.getNumeClient(), rezervare1.getNrLocuri(), rezervare1.getCursa().getId())).orElse(null);
    }


    @Override
    public void makeReservation(Long idCursa, String numeClient, Integer nrLocuri) throws AppException {

        int locuriDisponibile = 18 - locuriRepo.findUltimulLoc(idCursa);
        if (nrLocuri > locuriDisponibile) {
            throw new AppException("Nu sunt suficiente locuri disponibile");
        }

        // Create a new reservation
        Rezervare rezervare = new Rezervare(null, numeClient, nrLocuri, cursaRepo.findOne(idCursa).get());
        ValidatorRezervare validatorRezervare = new ValidatorRezervare();
        try {
            validatorRezervare.validate(rezervare);
            notifyClients();

        } catch (ValidationException e) {
            throw new AppException(e.getMessage());
        }

        rezervare = rezervareRepo.save(rezervare);

        notifyClients();
    }

    private void notifyClients(){
        ExecutorService executor = Executors.newFixedThreadPool(defaultThreadsNo);
        for(var client : loggedClients.values()){
            if(client == null)
                continue;
            executor.execute(client::reservationUpdate);


        }
    }
}

File: mpp-firma-transport-c#/FirmaServer/FirmaServices.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using System.Threading.Tasks;
using FirmaModel1;
using FirmaPersistence3;
using FirmaService;
using Utils;

namespace FirmaServer
{
    public class FirmaServices : IFirmaServices
    {
        private IOficiuRepository oficiuRepo;
        private ILocuriRepository locuriRepo;
        private ICursaRepository cursaRepo;
        private IRezervareRepository rezervareRepo;
        private Dictionary<string, IFirmaObserver> _loggedClients;
        private readonly int defaultThreads = 5;
        private readonly object _lock = new object();

        public FirmaServices(IOficiuRepository oficiuRepo, ILocuriRepository locuriRepo, ICursaRepository cursaRepo, IRezervareRepository rezervareRepo)
        {
            this.oficiuRepo = oficiuRepo;
            this.locuriRepo = locuriRepo;
            this.cursaRepo = cursaRepo;
            this.rezervareRepo = rezervareRepo;
            _loggedClients = new Dictionary<string, IFirmaObserver>();
        }



        public Optional<Oficiu> login(string username, string password, IFirmaObserver client)
        {
            Console.WriteLine("Login request");
            lock (_lock)
            {
                var utilizatorOpt = oficiuRepo.FindByUserPassword(username, password);
                if (utilizatorOpt.HasValue)
                {
                    var utilizator = utilizatorOpt.Value;
                    var parolaDB = utilizator.Password;

                    if (_loggedClients.ContainsKey(username))
                    {
                        throw new AppException("Utilizator logat deja");
                    }
                    else
                    {
                        _loggedClients[username] = client;
                        return Optional<Oficiu>.Of(utilizator);
                    }
                }

                return Optional<Oficiu>.Empty();
            }
        }

        public void logout(string username)
        {
            lock (_lock)
            {
                _loggedClients.Remove(username);
            }
        }

        public List<CursaDTO> getAllCurse()
        {
            try
            {
                // Get all curse from cursaRepo
                var allCurse = cursaRepo.FindAll();

                // Transform curse into DTOs
                var cursaDTOs = allCurse.Select(cursa =>
                {
                    var locuriDisponibile = 18 - locuriRepo.findultimulLoc(cursa.Id);
                    return new CursaDTO(cursa.Id, cursa.Destinatie, cursa.Data, locuriDisponibile);
                }).ToList();

                return cursaDTOs;
            }
            catch (Exception ex)
            {
                // Handle exception if needed
                // You might log the exception or throw a custom exception
                throw new AppException("Error occurred while getting all curse.");
            }
        }

        public List<CursaDTO> getCursaByDestinatieData(string destinatie, string data)
        {
            try
            {
                // Find Cursa by destinatie and data
                Optional<Cursa> optionalCursa = cursaRepo.FindByDestinatieData(destinatie, data);

                if (optionalCursa.HasValue)
                {
                    Cursa cursa = optionalCursa.Value;
                    int locuriDisponibile = 18 - locuriRepo.findultimulLoc(cursa.Id);
                    CursaDTO cursaDTO = new CursaDTO(cursa.Id, cursa.Destinatie, cursa.Data, locuriDisponibile);
                    List<CursaDTO> cursaDTOs = new List<CursaDTO> ();
                    cursaDTOs.Add(cursaDTO);
                    return cursaDTOs; // Return a list containing the DTO
                }
                else
                {
                    return new List<CursaDTO>(); // Return an empty list if no matching record was found
                }
            }
            catch (Exception ex)
            {
                // Handle exception if needed
                // You might log the exception or throw a custom exception
                throw new AppException("Error occurred while getting cursa by destinatie and data.");
            }
        }

        public List<LocDTO> getLocuriDisponibile(long idCursa)
        {
            return locuriRepo.findByCursa(idCursa);


        }

        public RezervareDTO getRezervareById(long rezervareId)
        {
            Rezervare rezervare = rezervareRepo.FindOne(rezervareId).Value;

            if (rezervare != null)
            {
                return new RezervareDTO(rezervare.Id, rezervare.NumeClient, rezervare.NrLocuri, rezervare.Cursa.Id);
            }
            else
            {
                return null;
            }
        }

        public void makeReservation(long idCursa, String numeClient, int nrLocuri)
        {
            try
            {
                int locuriDisponibile = 18 - locuriRepo.findultimulLoc(idCursa);

                if (nrLocuri > locuriDisponibile)
                {
                    throw new AppException("Nu sunt suficiente locuri disponibile");
                }

                // Create a new reservation
                Optional<Cursa> cursa = cursaRepo.FindOne(idCursa);
                if (cursa.Value == null)
                {
                    throw new AppException("Cursa not found");
                }

                Rezervare rezervare = new Rezervare(numeClient, nrLocuri, cursa.Value);
                //rezervare.Id = rezervareRepo.FindAll().Last().Id + 1;



                // Save reservation
                Optional<Rezervare> rezervare1 = rezervareRepo.Save(rezervare);

                NotifyAllClients();
            }
            catch (ValidationException ex)
            {
                throw new AppException(ex.Message);
            }
            catch (AppException ex)
            {
                Console.WriteLine(ex.Message);
                // Handle other exceptions if needed
                throw new AppException("Error occurred while making reservation");
            }
        }

        private void NotifyAllClients()
        {
            Console.WriteLine("Notifying all clients ...");
            foreach (var client in _loggedClients.Values)
            {
                // client.ReservationUpdate();
                Task.Run(() => client.ReservationUpdate());
            }
        }


    }
}